stage1/init/common/mount.go

```diff
@@ -238,3 +242,150 @@ func ensureDestinationExists(source, destination string) error {
+ * limitation.
+ */
+func AppAddOneMount(p *stage1commontypes.Pod, ra *schema.RuntimeApp, sourcePath string, dstPath string, readOnly bool, enterCmd []string) {
+	/* Prepare a temporary playground that is not a shared mount */
+	playgroundMount, err := ioutil.TempDir("", "rkt.propagate.")
+	if err != nil {
+		log.FatalE("error creating temporary propagation directory", err)
+		os.Exit(1)
+	}
+	defer os.Remove(playgroundMount)
+
+	err = syscall.Mount(playgroundMount, playgroundMount, "bind", syscall.MS_BIND, "")
+	if err != nil {
+		log.FatalE("error mounting temporary directory", err)
+		os.Exit(1)
+	}
+	defer syscall.Unmount(playgroundMount, 0)
+
+	err = syscall.Mount("", playgroundMount, "none", syscall.MS_SLAVE, "")
+	if err != nil {
+		log.FatalE("error mounting temporary directory", err)
+		os.Exit(1)
+	}
+
+	/* Bind mount the source into the playground, possibly read-only */
+	mountTmp := filepath.Join(playgroundMount, "mount")
+	if err := os.MkdirAll(mountTmp, 0700); err != nil {
+		log.FatalE("error creating temporary mount directory", err)
+		os.Exit(1)
+	}
+	defer os.Remove(mountTmp)
+
+	err = syscall.Mount(sourcePath, mountTmp, "bind", syscall.MS_BIND, "")
+	if err != nil {
+		log.FatalE("error mounting temporary directory", err)
+		os.Exit(1)
+	}
+	defer syscall.Unmount(mountTmp, 0)
+
+	if readOnly {
+		err = syscall.Mount("", mountTmp, "bind", syscall.MS_REMOUNT|syscall.MS_RDONLY|syscall.MS_BIND, "")
+		if err != nil {
+			log.FatalE("error remounting temporary mount directory read-only", err)
+			os.Exit(1)
+		}
+	}
+
+	/* Now that the bind mount has the correct attributes (RO or RW), move
+	 * it to the propagation directory prepared by systemd-nspawn */
+	mountOutside := filepath.Join("/run/systemd/nspawn/propagate/", "rkt-"+p.UUID.String(), "rkt.mount")
+	mountInside := filepath.Join("/run/systemd/nspawn/incoming/", filepath.Base(mountOutside))
+	/* TODO(Alban): instead of always using mkdir(), we should check if the
+	 * source is a file or a directory, and use mkdir() or touch() accordingly. */
+	if err := os.MkdirAll(mountOutside, 0700); err != nil {
+		log.FatalE("error creating temporary mount directory", err)
+		os.Exit(1)
+	}
+	defer os.Remove(mountOutside)
+
+	err = syscall.Mount(mountTmp, mountOutside, "", syscall.MS_MOVE, "")
+	if err != nil {
+		log.FatalE("error moving mount directory", err)
+		os.Exit(1)
+	}
+	defer syscall.Unmount(mountOutside, 0)
+
+	mountDst := filepath.Join("/opt/stage2", ra.Name.String(), "rootfs", dstPath)
+	mountDstOutside := filepath.Join(p.Root, "stage1/rootfs", mountDst)
+	/* TODO(Alban): ditto, possibly use touch() instead. */
+	if err := os.MkdirAll(mountDstOutside, 0700); err != nil {
+		log.FatalE("error creating destination directory", err)
+		os.Exit(1)
+	}
+
+	/* Finally move the bind mount at the correct place inside the
+	 * container. */
+	args := enterCmd
+	args = append(args, "/bin/mount", "--move", mountInside, mountDst)
```

Comment:
We need to enter into the stage1 namespace, we can't implement that in go so we just enter with the entrypoint and call mount. We could do a custom thing that calls the mount syscall but just calling `/bin/mount` is easier